akka-http-core/src/main/java/akka/http/javadsl/model/headers/ModeledCustomHeaderFactory.java

```diff
@@ -0,0 +1,43 @@
+package akka.http.javadsl.model.headers;
+
+import akka.http.javadsl.model.HttpHeader;
+import akka.util.Helpers;
+
+import java.util.Optional;
+
+/**
+ * Companion class for the {@link ModeledCustomHeader} class. It offers methods to create {@link ModeledCustomHeader}
+ * from {@link String} or {@link HttpHeader}.
+ */
+public abstract class ModeledCustomHeaderFactory<H extends ModeledCustomHeader> {
+
+  public abstract String name();
+
+  public String lowercaseName() {
+    return Helpers.toRootLowerCase(name());
+  }
+
+  /**
+   * Parses the value checking that the format is correct.
+   * It may throw if value is not correct
+   */
+  protected abstract H parse(final String value);
+
+  /**
+   * Creates a new {@code ModeledCustomHeader} from the value checking that the format is correct.
+   * It may throw if value is not correct
+   */
+  public H create(final String value) {
+    return parse(value);
+  }
+
+  /**
+   * Transforms an {@code HttpHeader} to this {@code ModeledCustomHeader} if the name and value are correct.
+   */
+  public Optional<H> from(final HttpHeader header) {
+    if (header.lowercaseName().equals(lowercaseName())) {
+      return Optional.of(parse(header.value()));
```

Comment:
This can throw; the return type implies that if we can't parse we would get a None yet we would not.

Document this behaviour, I think it's OK but should be explained.
Since a malformed header then will blow up and get logged, if just not present we nicely give a None, that's fine.